<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Connect-the-sql" data-toc-modified-id="Connect-the-sql-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Connect the sql</a></span></li><li><span><a href="#Know-your-database" data-toc-modified-id="Know-your-database-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Know your database</a></span></li><li><span><a href="#Read-dat-files-of-database" data-toc-modified-id="Read-dat-files-of-database-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Read dat files of database</a></span><ul class="toc-item"><li><span><a href="#Get-pandas-dataframes" data-toc-modified-id="Get-pandas-dataframes-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Get pandas dataframes</a></span></li></ul></li><li><span><a href="#Section-2" data-toc-modified-id="Section-2-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Section 2</a></span><ul class="toc-item"><li><span><a href="#qn1" data-toc-modified-id="qn1-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>qn1</a></span></li><li><span><a href="#qn2" data-toc-modified-id="qn2-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>qn2</a></span></li><li><span><a href="#qn3" data-toc-modified-id="qn3-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>qn3</a></span></li><li><span><a href="#qn" data-toc-modified-id="qn-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>qn</a></span></li><li><span><a href="#qn" data-toc-modified-id="qn-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>qn</a></span></li></ul></li></ul></div>

# Connect the sql

In [1]:
import numpy as np
import pandas as pd
import os
import yaml

with open( os.path.expanduser('~') + "/.postgres_conf.yml", 'r') as stream:
    try:
        yaml_dict = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

pw = yaml_dict['password']
port = yaml_dict['port']

In [2]:
%load_ext sql

In [3]:
%sql postgres://postgres:$pw@localhost:$port/dvdrental

'Connected: postgres@dvdrental'

# Know your database

In [4]:
%%sql
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE'
AND TABLE_CATALOG='dvdrental'
and TABLE_NAME  not like 'pg_%'
and TABLE_NAME  not like 'sql_%'
-- order by table_name

 * postgres://postgres:***@localhost:5433/dvdrental
15 rows affected.


[('staff',),
 ('category',),
 ('film_category',),
 ('country',),
 ('actor',),
 ('language',),
 ('inventory',),
 ('payment',),
 ('rental',),
 ('city',),
 ('store',),
 ('film',),
 ('address',),
 ('film_actor',),
 ('customer',)]

In [5]:
%%sql
select column_name, data_type, character_maximum_length
    from INFORMATION_SCHEMA.COLUMNS 
    where table_name = 'customer';

 * postgres://postgres:***@localhost:5433/dvdrental
10 rows affected.


[('customer_id', 'integer', None),
 ('store_id', 'smallint', None),
 ('first_name', 'character varying', 45),
 ('last_name', 'character varying', 45),
 ('email', 'character varying', 50),
 ('address_id', 'smallint', None),
 ('activebool', 'boolean', None),
 ('create_date', 'date', None),
 ('last_update', 'timestamp without time zone', None),
 ('active', 'integer', None)]

In [6]:
%%sql
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE'
AND TABLE_CATALOG='dvdrental'
and TABLE_NAME  not like 'pg_%'
and TABLE_NAME  not like 'sql_%'
-- order by table_name

 * postgres://postgres:***@localhost:5433/dvdrental
15 rows affected.


[('staff',),
 ('category',),
 ('film_category',),
 ('country',),
 ('actor',),
 ('language',),
 ('inventory',),
 ('payment',),
 ('rental',),
 ('city',),
 ('store',),
 ('film',),
 ('address',),
 ('film_actor',),
 ('customer',)]

In [7]:
%%sql
-- select * from staff limit 2; -- This gives error in jupyter notebook
select * from actor limit 2;

 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.


[(1, 'Penelope', 'Guiness', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (2, 'Nick', 'Wahlberg', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000))]

# Read dat files of database

In [8]:
!ls ../data/dvdrental/

2163.dat    2167.dat    2169.dat    2173.dat    2177.dat    2181.dat    2185.dat    2189.dat    toc.dat
2165.dat    2168.dat    2171.dat    2175.dat    2179.dat    2183.dat    2187.dat    restore.sql


In [9]:
!head ../data/dvdrental/2163.dat

1	Penelope	Guiness	2013-05-26 14:47:57.62
2	Nick	Wahlberg	2013-05-26 14:47:57.62
3	Ed	Chase	2013-05-26 14:47:57.62
4	Jennifer	Davis	2013-05-26 14:47:57.62
5	Johnny	Lollobrigida	2013-05-26 14:47:57.62
6	Bette	Nicholson	2013-05-26 14:47:57.62
7	Grace	Mostel	2013-05-26 14:47:57.62
8	Matthew	Johansson	2013-05-26 14:47:57.62
9	Joe	Swank	2013-05-26 14:47:57.62
10	Christian	Gable	2013-05-26 14:47:57.62


In [10]:
%%bash

for f in `ls ../data/dvdrental/[^toc]*.dat`                                                                
   do
      echo
      echo  "File: $f"; 
      echo "==================================================================="
      head -2  "$f";
done;


File: ../data/dvdrental/2163.dat
1	Penelope	Guiness	2013-05-26 14:47:57.62
2	Nick	Wahlberg	2013-05-26 14:47:57.62

File: ../data/dvdrental/2165.dat
1	Action	2006-02-15 09:46:27
2	Animation	2006-02-15 09:46:27

File: ../data/dvdrental/2167.dat
133	Chamber Italian	A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria	2006	1	7	4.99	117	14.99	NC-17	2013-05-26 14:50:58.951	{Trailers}	'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5
384	Grosse Wonderful	A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia	2006	1	5	4.99	49	19.99	R	2013-05-26 14:50:58.951	{"Behind the Scenes"}	'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2

File: ../data/dvdrental/2168.dat
1	1	2006-02-15 10:05:03
1	23	2006-02-15 10:05:03

File: ../data/dvdrental/2169.dat
1	6	2006-02-15 10:07:09
2	11	2006-02-15 10:07:09

File: ../data/dvdrental/2171.dat

## Get pandas dataframes
- 2187.dat staff
- 2165.dat category
- 2179.dat film_category
- 2175.dat country
- 2163.dat actor
- 2181.dat language
- 2179.dat inventory
- 2183.dat payment
- 2185.dat rental
- 2173.dat city
- 2179.dat store
- 2167.dat  film
- 2171.dat address
- 2168.dat film_actor
- 2177.dat customer

In [11]:
tables = ['staff', 'category', 'film_category', 'country', 'actor',
          'language', 'inventory', 'payment', 'rental', 'city',
          'store', 'film', 'address', 'film_actor', 'customer']

staff = tables[0]
category = tables[1]
film_category = tables[2]
country = tables[3]
actor = tables[4]
language = tables[5]
inventory = tables[6]
payment = tables[7]
rental = tables[8]
city = tables[9]
store = tables[10]
film = tables[11]
address = tables[12]
film_actor = tables[13]
customer = tables[14]

In [12]:
# staff = %sql select * from $staff;
# staff = customer.DataFrame()


In [13]:
category = %sql select * from $category;
category = category.DataFrame()

 * postgres://postgres:***@localhost:5433/dvdrental
16 rows affected.


In [14]:
# staff = %sql select * from $staff;  # this fails

film_category = %sql select * from $film_category;
country = %sql select * from $country;
actor = %sql select * from $actor;
language = %sql select * from $language;
inventory = %sql select * from $inventory;
payment = %sql select * from $payment;
rental = %sql select * from $rental;
city = %sql select * from $city;
store = %sql select * from $store;
film = %sql select * from $film;
address = %sql select * from $address;
film_actor = %sql select * from $film_actor;
customer = %sql select * from $customer;

 * postgres://postgres:***@localhost:5433/dvdrental
1000 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
109 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
200 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
6 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
4581 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
14596 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
16044 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
600 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
1000 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
603 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
5462 rows affected.
 * postgres://postgres:***@localhost:5433/dvdrental
599 rows affected.


In [15]:
film_category = film_category.DataFrame()
country = country.DataFrame()
actor = actor.DataFrame()
language = language.DataFrame()
inventory = inventory.DataFrame()
payment = payment.DataFrame()
rental = rental.DataFrame()
city = city.DataFrame()
store = store.DataFrame()
film = film.DataFrame()
address = address.DataFrame()
film_actor = film_actor.DataFrame()
customer = customer.DataFrame()

In [16]:
# I got memory error for city table of file 2173.dat
# I will manually read the datafile using pandas

In [17]:
# I can not open staff table in jupyter notebook, but I can open this in
# Query tool of pgAdmin.

"""
select column_name from INFORMATION_SCHEMA.COLUMNS 
where table_name = 'customer';


Gives:

"staff_id"
"first_name"
"last_name"
"address_id"
"email"
"store_id"
"active"
"username"
"password"
"last_update"
"picture"


"""

'\nselect column_name from INFORMATION_SCHEMA.COLUMNS \nwhere table_name = \'customer\';\n\n\nGives:\n\n"staff_id"\n"first_name"\n"last_name"\n"address_id"\n"email"\n"store_id"\n"active"\n"username"\n"password"\n"last_update"\n"picture"\n\n\n'

In [18]:
cols = ['staff_id', 'first_name', 'last_name', 'address_id', 'email',
        'store_id', 'active', 'username', 'password', 'last_update',
        'picture']

In [19]:
!head ../data/dvdrental/2187.dat

1	Mike	Hillyer	3	Mike.Hillyer@sakilastaff.com	1	t	Mike	8cb2237d0679ca88db6464eac60da96345513964	2006-05-16 16:13:11.79328	\\x89504e470d0a5a0a
2	Jon	Stephens	4	Jon.Stephens@sakilastaff.com	2	t	Jon	8cb2237d0679ca88db6464eac60da96345513964	2006-05-16 16:13:11.79328	\N
\.




In [22]:
staff = pd.read_csv('../data/dvdrental/2187.dat', sep=r'\t', header=None, engine='python')
staff.columns = cols
staff = staff.head(2)
staff['active'] = True
print(staff.shape)
staff.head()

# looking at pgAdmin this table has only two rows and last column is nulls.
# active column t mean true.
# 

(2, 11)


staff_id first_name last_name  address_id                         email  \
0        1       Mike   Hillyer         3.0  Mike.Hillyer@sakilastaff.com   
1        2        Jon  Stephens         4.0  Jon.Stephens@sakilastaff.com   

   store_id  active username                                  password  \
0       1.0    True     Mike  8cb2237d0679ca88db6464eac60da96345513964   
1       2.0    True      Jon  8cb2237d0679ca88db6464eac60da96345513964   

                 last_update              picture  
0  2006-05-16 16:13:11.79328  \\x89504e470d0a5a0a  
1  2006-05-16 16:13:11.79328                   \N

In [57]:
df_tables = [staff, category, film_category, country, actor,
            language, inventory, payment, rental, city,
            store, film, address, film_actor, customer]

In [58]:
all_columns = [df.columns.tolist() for df in df_tables]
all_columns[0]

['staff_id',
 'first_name',
 'last_name',
 'address_id',
 'email',
 'store_id',
 'active',
 'username',
 'password',
 'last_update',
 'picture']

In [65]:
df_tables_cols = pd.DataFrame(all_columns).T.fillna('')

df_tables_cols.columns = tables
df_tables_cols

staff     category film_category      country        actor  \
0      staff_id  category_id       film_id   country_id     actor_id   
1    first_name         name   category_id      country   first_name   
2     last_name  last_update   last_update  last_update    last_name   
3    address_id                                          last_update   
4         email                                                        
5      store_id                                                        
6        active                                                        
7      username                                                        
8      password                                                        
9   last_update                                                        
10      picture                                                        
11                                                                     
12                                                                     

       language     inventory       payment        rental         city  \
0   language_id  inventory_id    payment_id     rental_id      city_id   
1          name       film_id   customer_id   rental_date         city   
2   last_update      store_id      staff_id  inventory_id   country_id   
3                 last_update     rental_id   customer_id  last_update   
4                                    amount   return_date                
5                              payment_date      staff_id                
6                                             last_update                
7                                                                        
8                                                                        
9                                                                        
10                                                                       
11                                                                       
12                                                                       

               store              film      address   film_actor     customer  
0           store_id           film_id   address_id     actor_id  customer_id  
1   manager_staff_id             title      address      film_id     store_id  
2         address_id       description     address2  last_update   first_name  
3        last_update      release_year     district                 last_name  
4                          language_id      city_id                     email  
5                      rental_duration  postal_code                address_id  
6                          rental_rate        phone                activebool  
7                               length  last_update               create_date  
8                     replacement_cost                            last_update  
9                               rating                                 active  
10                         last_update                                         
11                    special_features                                         
12                            fulltext

# Section 2

## qn1
How many payment transactions were greater than $5 ?

In [21]:
%%sql
select * from customer limit 5;

 * postgres://postgres:***@localhost:5433/dvdrental
5 rows affected.


[(524, 1, 'Jared', 'Ely', 'jared.ely@sakilacustomer.org', 530, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (1, 1, 'Mary', 'Smith', 'mary.smith@sakilacustomer.org', 5, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (2, 1, 'Patricia', 'Johnson', 'patricia.johnson@sakilacustomer.org', 6, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (3, 1, 'Linda', 'Williams', 'linda.williams@sakilacustomer.org', 7, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1),
 (4, 2, 'Barbara', 'Jones', 'barbara.jones@sakilacustomer.org', 8, True, datetime.date(2006, 2, 14), datetime.datetime(2013, 5, 26, 14, 49, 45, 738000), 1)]

In [24]:
%%sql
select * from payment limit 2;

 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.


[(17503, 341, 2, 1520, Decimal('7.99'), datetime.datetime(2007, 2, 15, 22, 25, 46, 996577)),
 (17504, 341, 1, 1778, Decimal('1.99'), datetime.datetime(2007, 2, 16, 17, 23, 14, 996577))]

In [25]:
%%sql
select count(*) from payment
where amount > 5;

 * postgres://postgres:***@localhost:5433/dvdrental
1 rows affected.


[(3618,)]

In [31]:
payment.query('amount > 5').size # ignores nulls

21708

In [32]:
payment.query('amount > 5').shape[0]

3618

## qn2
How many actors have a first name that starts with the letter P?


In [33]:
%%sql
select * from actor limit 5;

 * postgres://postgres:***@localhost:5433/dvdrental
5 rows affected.


[(1, 'Penelope', 'Guiness', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (2, 'Nick', 'Wahlberg', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (3, 'Ed', 'Chase', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (4, 'Jennifer', 'Davis', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000)),
 (5, 'Johnny', 'Lollobrigida', datetime.datetime(2013, 5, 26, 14, 47, 57, 620000))]

In [34]:
%%sql
select count(*) from actor
where first_name like 'P%';

 * postgres://postgres:***@localhost:5433/dvdrental
1 rows affected.


[(5,)]

In [37]:
actor[actor.first_name.str.startswith('P')].shape

(5, 4)

In [38]:
actor[actor.first_name.str.startswith('P')]

actor_id first_name last_name             last_update
0           1   Penelope   Guiness 2013-05-26 14:47:57.620
45         46     Parker  Goldberg 2013-05-26 14:47:57.620
53         54   Penelope   Pinkett 2013-05-26 14:47:57.620
104       104   Penelope    Cronyn 2013-05-26 14:47:57.620
120       120   Penelope    Monroe 2013-05-26 14:47:57.620

## qn3
How many unique districts are our customers from?


In [66]:
df_tables_cols

staff     category film_category      country        actor  \
0      staff_id  category_id       film_id   country_id     actor_id   
1    first_name         name   category_id      country   first_name   
2     last_name  last_update   last_update  last_update    last_name   
3    address_id                                          last_update   
4         email                                                        
5      store_id                                                        
6        active                                                        
7      username                                                        
8      password                                                        
9   last_update                                                        
10      picture                                                        
11                                                                     
12                                                                     

       language     inventory       payment        rental         city  \
0   language_id  inventory_id    payment_id     rental_id      city_id   
1          name       film_id   customer_id   rental_date         city   
2   last_update      store_id      staff_id  inventory_id   country_id   
3                 last_update     rental_id   customer_id  last_update   
4                                    amount   return_date                
5                              payment_date      staff_id                
6                                             last_update                
7                                                                        
8                                                                        
9                                                                        
10                                                                       
11                                                                       
12                                                                       

               store              film      address   film_actor     customer  
0           store_id           film_id   address_id     actor_id  customer_id  
1   manager_staff_id             title      address      film_id     store_id  
2         address_id       description     address2  last_update   first_name  
3        last_update      release_year     district                 last_name  
4                          language_id      city_id                     email  
5                      rental_duration  postal_code                address_id  
6                          rental_rate        phone                activebool  
7                               length  last_update               create_date  
8                     replacement_cost                            last_update  
9                               rating                                 active  
10                         last_update                                         
11                    special_features                                         
12                            fulltext

In [68]:
%%sql
select * from address limit 2;

 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.


[(1, '47 MySakila Drive', None, 'Alberta', 300, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30)),
 (2, '28 MySQL Boulevard', None, 'QLD', 576, '', '', datetime.datetime(2006, 2, 15, 9, 45, 30))]

In [69]:
%%sql
select count(distinct(district)) from address;

 * postgres://postgres:***@localhost:5433/dvdrental
1 rows affected.


[(378,)]

In [70]:
address.district.nunique()

378

## qn
How many films have a rating of R and a replacement cost between `$5` and `$15` ?

In [71]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [75]:
%%sql
select count(*) from film
where rating = 'R' and replacement_cost between 5 and 15;

 * postgres://postgres:***@localhost:5433/dvdrental
1 rows affected.


[(52,)]

In [77]:
film.query("rating == 'R' and 5 <= replacement_cost <= 15 ").shape

(52, 13)

## qn
How many films have the word Truman somewhere in the title?



In [78]:
%%sql
select * from film limit 2;

 * postgres://postgres:***@localhost:5433/dvdrental
2 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2")]

In [79]:
%%sql
select count(*) from film
where title like '%Truman%';

 * postgres://postgres:***@localhost:5433/dvdrental
1 rows affected.


[(5,)]

In [80]:
%%sql
select film_id, title from film
where title like '%Truman%';

 * postgres://postgres:***@localhost:5433/dvdrental
5 rows affected.


[(6, 'Agent Truman'),
 (537, 'Lover Truman'),
 (579, 'Minds Truman'),
 (613, 'Mystic Truman'),
 (915, 'Truman Crazy')]

In [84]:
film[film.title.str.contains('Truman')].iloc[:,:2]

film_id          title
9          6   Agent Truman
538      537   Lover Truman
579      579   Minds Truman
613      613  Mystic Truman
914      915   Truman Crazy

In [ ]:
## qn
